pkg/volume/csi/nodeinfomanager/nodeinfomanager.go

```diff
@@ -86,11 +92,161 @@ type Interface interface {
 // NewNodeInfoManager initializes nodeInfoManager
 func NewNodeInfoManager(
 	nodeName types.NodeName,
-	volumeHost volume.VolumeHost) Interface {
+	volumeHost volume.VolumeHost,
+	migratedDrivers map[string](func() bool)) Interface {
 	return &nodeInfoManager{
-		nodeName:   nodeName,
-		volumeHost: volumeHost,
+		nodeName:        nodeName,
+		volumeHost:      volumeHost,
+		migratedDrivers: migratedDrivers,
+	}
+}
+
+// InitializeMigratedDrivers will create CSINodeInfo objects for all drivers that are to
+// be migrated. It will initialize most fields with dummy values. The CSIDriverInfo object
+// will be initialized with DriverName, ProxyInTreeVolumePlugin, and Available.
+func (nim *nodeInfoManager) InitializeMigratedDrivers() error {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.CSINodeInfo) ||
+		!utilfeature.DefaultFeatureGate.Enabled(features.CSIMigration) {
+		return fmt.Errorf("initializeMigratedDrivers called but CSINodeInfo or CSIMigration is disabled")
+	}
+
+	var updateErrs []error
+
+	csiKubeClient := nim.volumeHost.GetCSIClient()
+	if csiKubeClient == nil {
+		return fmt.Errorf("error getting CSI client")
+	}
+
+	// TODO(dyzz): This should probably retry forever!!
+	// TODO(TODO): Change this into reconcile pattern
+	// Loop to initialize Specs
+	err := wait.ExponentialBackoff(updateBackoff, func() (bool, error) {
+		nodeInfo, err := csiKubeClient.CsiV1alpha1().CSINodeInfos().Get(string(nim.nodeName), metav1.GetOptions{})
+		if nodeInfo == nil || errors.IsNotFound(err) {
+			nodeInfo, err = nim.CreateCSINodeInfo()
+		}
+		if err != nil {
+			updateErrs = append(updateErrs, err)
+			return false, nil
+		}
+		err = nim.tryInitializeMigratedDriverSpecs(nodeInfo)
+		if err != nil {
+			updateErrs = append(updateErrs, err)
+			return false, nil
+		}
+		return true, nil
+
+	})
+
+	// Loop to initialize Status
+	// TODO(TODO) We can remove this once nodeinfomanager is refactored into a reconcile pattern because
+	// then we can just add the specs in the above loop and the reconciler will fill in the statuses based
+	// on that.
+	err = wait.ExponentialBackoff(updateBackoff, func() (bool, error) {
```

Comment:
same comment here about not retrying inside InitializeMigratedDrivers and depending on the caller to re-call us

Reply:
can also use the updated object that was the output of the previous step instead of looking it up again